regnety/dataset/imagenet.py

```diff
@@ -0,0 +1,268 @@
+        augment_fn: function to apply to dataset after loading raw TFrecords
+                    'default' implies following augmentations will be applied:
+                    - Random sized crop (train only)
+                    - Scale and center crop
+                    - RandAugment (TODO)
+        num_classes: number of classes
+        randaugment: True of RandAugment is to be applied
+
+    Class attributes:
+        TFRECS_FORMAT: Expected format of TFRecords stored.
+        _MEAN: Channel wise mean for ImageNet
+        _STD: Channel wise standard deviation for ImageNet
+
+    """
+
+    TFRECS_FORMAT = {
+        "image": tf.io.FixedLenFeature([], tf.string),
+        # "height": tf.io.FixedLenFeature([], tf.int64),
+        # "width": tf.io.FixedLenFeature([], tf.int64),
+        "filename": tf.io.FixedLenFeature([], tf.string),
+        "label": tf.io.FixedLenFeature([], tf.int64),
+        "synset": tf.io.FixedLenFeature([], tf.string),
+    }
+
+    _MEAN = tf.constant([0.485, 0.456, 0.406])
+    _STD = tf.constant([0.229, 0.224, 0.225])
+
+    def __init__(
+        self,
+        tfrecs_filepath=None,
+        batch_size=128,
+        image_size=224,
+        augment_fn="default",
+        num_classes=10,
+        randaugment=True,
+    ):
+
+        if tfrecs_filepath is None:
+            raise ValueError("List of TFrecords paths cannot be None")
+        self.tfrecs_filepath = tfrecs_filepath
+        self.batch_size = batch_size
+        self.image_size = image_size
+        self.augment_fn = augment_fn
+        self.num_classes = num_classes
+        self.randaugment = randaugment
+        if self.randaugment:
+            self.augmenter = RandAugment(magnitude=5)
+
+    def decode_example(self, example):
+        """Decodes an example to its individual attributes
+
+        Args:
+            example: A TFRecord dataset example.
+
+        Returns:
+            Dict containing attributes from a single example. Follows
+            the same names as TFRECORDS_FORMAT.
+        """
+        image = tf.cast(tf.io.decode_jpeg(example["image"]), tf.float32)
+        height = 512
+        width = 512
+        filename = example["filename"]
+        label = example["label"]
+        synset = example["synset"]
+        return {
+            "image": image,
+            "height": height,
+            "width": width,
+            "filename": filename,
+            "label": label,
+            "synset": synset,
+        }
+
+    def _read_tfrecs(self):
+        """Function for reading and loading TFRecords into a tf.data.Dataset.
+
+        Returns:
+            A tf.data.Dataset
+        """
+        ds = tf.data.TFRecordDataset(self.tfrecs_filepath)
+        ds = ds.map(
+            lambda example: tf.io.parse_example(example, ImageNet.TFRECS_FORMAT)
+        )
+        ds = ds.map(lambda example: self.decode_example(example))
+        return ds
+
+    def _scale_and_center_crop(self, image, h, w, scale_size, final_size):
```

Comment:
A downside of this approach is you are always cropping non-square images. What do you think of setting the *longer* side to `scale_size`, and adding black padding to the shorter side? You could remove most of your logic and achieve this by simply calling:

```python
square_scaled_image = tf.image.resize_with_pad(image, scale_size, scale_size)
return tf.image.central_crop(tf.cast(final_size, tf.float32) / scale_size)
```

Reply:
Sure, that sounds good